### PR TITLE
Use reference to pass ufc_form

### PR DESCRIPTION
**Pull request by chrisrichardson**
Stop using `shared_ptr` where it is no longer needed for ufc objects passed into dolfin.

### CODE REVIEW

**Comment by michalhabera** [comment]
So this makes cpp user hands dirty with raw pointer. What is the argument for that?
You could have passed by reference and still use smart pointers.

**Comment by chrisrichardson** [comment]
There is always a raw pointer from UFC. The alternative is `std::unique_ptr<ufc_foo, decltype(std::free)*>(ufc_foo_create(), std::free);`. So, you still need to tell the "smart" pointer how to destroy the malloc-created object. I think it is better to just use `free`. We can wrap all this up in a utility function, perhaps, later.

File: cpp/demo/hyperelasticity/main.cpp [hunk 1]
```diff
@@ -170,15 +170,15 @@ int main(int argc, char* argv[])
...
-
-  auto ufc_map = std::shared_ptr<ufc_dofmap>(space->dofmap(), free);
+  ufc_function_space* space = hyperelasticity_functionspace_create();
+  ufc_dofmap* ufc_map = space->dofmap();
   auto V = std::make_shared<function::FunctionSpace>(
       mesh,
       std::make_shared<fem::FiniteElement>(
           std::shared_ptr<ufc_finite_element>(space->element(), free)),
       std::make_shared<fem::DofMap>(*ufc_map, *mesh));
+  std::free(ufc_map);
```
Comment (garth-wells) [inline cpp/demo/hyperelasticity/main.cpp]: Is this correct? Does `space->dofmap()` create a dofmap, or just point to the dofmap owned by `space`?

Comment (chrisrichardson) [reply cpp/demo/hyperelasticity/main.cpp]: It creates it. I guess it should have create in the name, but that's an ffc/ufc issue.

**Comment by michalhabera** [comment]
Ok, just some ufc objects are handled with smart (ufc_finite_element) and some with raw pointers, which is not consistent now. Would it make sense to keep just one approach?

**Comment by chrisrichardson** [comment]
@michalhabera - it still makes sense to use `shared_ptr` on `ufc_finite_element`, and `ufc_coordinate_mapping` because they are stored in the appropriate dolfin classes (a situation I would also like to change, but that will take longer).

**Pull request merged**